### PR TITLE
Add example for how to store and restore a network.

### PR DESCRIPTION
**Pull request by heplesser**
Here comes my example for how to dump and load in NEST, inspired by #1828. Since the example is code-heavy, I provide it here as an example, but we could consider a guide as well. The resulting figure is below.

Giving it high priority because this topic has waited for documentation for a long time.

### Note
- **This uses the new random features already, so it should not be merged before #1549.**
- Because it builds on #1549, it comprises at the moment a huge number of files. The only relevant file is `pynest/examples/store_restore_network.py`.

dump_load_demo

### CODE REVIEW

**Comment by jougs** [comment]
@heplesser: many thanks for this valuable addition. The figures look indeed very nice and instructive.

Would it be OK for you if I postponed my review until after the merge of #1549? Right now, it is a bit hard to see what's what and I don't want to waste any time on reviewing the wrong code.

**Comment by heplesser** [comment]
@jougs No problem, #1549 should indeed be merged first. I will change priority to Normal on this one.

**Comment by clinssen** [comment]
I would suggest to either use a neuron with delta synapses, or to also store the state of I_ex_, I_in_, dI_ex_ and dI_in_ for each neuron, because they are readily read- and writeable.

**Comment by heplesser** [comment]
> I would suggest to either use a neuron with delta synapses, or to also store the state of I_ex_, I_in_, dI_ex_ and dI_in_ for each neuron, because they are readily read- and writeable.

Would you consider `iaf_psc_exp` and then storing `I_ex` and `I_in` suitable?

**Comment by clinssen** [comment]
Sure, that sounds like a good solution as well!

**Comment by YounesBouhadjar** [comment]
@heplesser this PR looks good to me, I can double-check it again when you introduce the changes @clinssen suggested.

**Comment by heplesser** [comment]
@YounesBouhadjar @clinssen With the new RNG interface merged now, this PR is ready for review :).

**Comment by YounesBouhadjar** [comment]
shouldn't we include this in the userdoc examples [here](https://github.com/nest/nest-simulator/tree/master/doc/userdoc/examples)?

**Comment by clinssen** [comment]
My earlier comment was not yet addressed: synaptic currents (I_ex in ``iaf_psc_exp``, I_ex and I_ex' in ``iaf_psc_alpha``) are not yet stored and restored.

**Comment by YounesBouhadjar** [comment]
@heplesser if you would like I can incorporate @clinssen suggestion in the code

**Comment by heplesser** [comment]
@clinssen I have now revised the example, also improving console output and figure design. I switched to `iaf_psc_delta` now, because I realised that it is currently not possible to obtain synaptic currents with `get()`, they are only available as recordables; they are also not settable. Changing that for all models would be a lot of manual work. We should introduce it when we replace the hand-written code with NESTML generated code. I don't think it would be a good idea to modify a single model just for the sake of this demo.

@YounesBouhadjar Could you help me with improving the layout of the figure? Font sizes are a bit of a mess and some more spacing is needed between the various axes. I think GridSpec can help with additional padding, but right now I have not enough time to figure it out myself ;).

**Comment by YounesBouhadjar** [comment]
@heplesser sure I can take care of this.

**Comment by YounesBouhadjar** [comment]
@heplesser what about this:
store_restore

**Comment by YounesBouhadjar** [comment]
We may need to improve the text in the caption.

**Comment by heplesser** [comment]
@YounesBouhadjar Much better! The "Continued Simulation" needs the x-axis labels, since there time runs from 1000 to 2000, while for all the reloaded ones below time starts again at 0.

I can fix the text in the box. Could you find out if there is an option to make the `text()` function handle the line breaks?

**Comment by YounesBouhadjar** [comment]
@heplesser Could you give me access right so I can push my changes? or how should we handle this?
I thought the function `text()` from matplotlib can handle line breaks, doesn't it?

**Comment by heplesser** [comment]
@YounesBouhadjar Can you push your version to your Github repo and then create a PR towards my personal repo and the dump_reload_branch in it? Then I will merge it into my branch and it will automatically appear here.

Concerning the `text()` function: I really know almost nothing about it, it might very well be that it does all we need automatically.

**Comment by YounesBouhadjar** [comment]
@heplesser should we get this done before the release of NEST3? the thing that still requires improvement in this PR is the text in the figure caption.

**Comment by heplesser** [comment]
@YounesBouhadjar @clinssen I have now revised the text in the figure. Could you both check if the PR is ready to be merged now?

File: pynest/examples/store_restore_network.py [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+        self.fig.suptitle("Storing and reloading a network simulation")
+        self.comment.set_axis_off()
+        self.comment.text(0, 1, textwrap.dedent("""
+            Storing, loading and continuing a simulation of a balanced E-I network
+            with STDP in excitatory synapses.
+
+            Top left: Raster plot of initial simulation for 1000ms (blue). Network state
+            (connections, membrane potential, synaptic weights) is stored at the end of
+            the initial simulation; synaptic currents are *not* stored.
```
Comment (clinssen) [inline pynest/examples/store_restore_network.py:227]:
```suggestion
            the initial simulation.
```

File: pynest/examples/store_restore_network.py [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+            Top right: Immediate continuation of the initial simulation from t=1000ms
+            to t=2000ms (orange) by calling Simulate(1000) again after storing the network.
+            This continues based on the full network state, including synaptic
```
Comment (clinssen) [inline pynest/examples/store_restore_network.py:231]:
```suggestion
            This continues based on the full network state, including spikes in transit.
```

File: pynest/examples/store_restore_network.py [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+            currents and spikes in transit.
```
Comment (clinssen) [inline pynest/examples/store_restore_network.py:232]:
```suggestion
```

File: pynest/examples/store_restore_network.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+    ein2 = EINetwork()
+    ein2.restore("ein_1000.pkl")
+    nest.Simulate(T_sim)
+    dplot.add_to_plot(ein2, lbl="Reloaded simulation (different seed)")
+
+    dplot.fig.savefig("store_restore_network.png")
+
+    input("Press ENTER to close figure!")
+
+    # NOTE: Comments below need updating after discussion about model to be used.
```
Comment (clinssen) [inline pynest/examples/store_restore_network.py:354]: Remove this comment block as information is already in the figure legend?

**Review by clinssen — CHANGES_REQUESTED** [review]

File: pynest/examples/store_restore_network.py (hunk 4 above)
Comment (heplesser) [reply pynest/examples/store_restore_network.py]: Done

File: pynest/examples/store_restore_network.py [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+
+This example shows how to store user-specified aspects of a network
+to file and how to later restore the network for further simulation.
+This may be used, e.g., to train weights in a network up to a certain
+point, store those weights and later perform diverse experiments on
+the same network using the stored weights.
+
+.. admonition:: Only user-specified aspects are stored
+
+   NEST does not support support storing the complete state of a simulation
```
Comment (YounesBouhadjar) [inline pynest/examples/store_restore_network.py:34]: leave only one `support` here

File: pynest/examples/store_restore_network.py [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+    ###############################################################################
+    # Create network from scratch and simulate 1s.
+    nest.SetKernelStatus({"local_num_threads": 4,
+                          "print_time": True})
+    ein = EINetwork()
+
+    print("*** Initial simulation ***")
+    ein.build()
+    nest.Simulate(T_sim)
+    dplot.add_to_plot(ein, lbl="Initial simuation")
```
Comment (YounesBouhadjar) [inline pynest/examples/store_restore_network.py:302]: typo in simulation

File: pynest/examples/store_restore_network.py [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+    # Store network state to file with state after 1s.
+    print("\n*** Storing simulation ...", end="", flush=True)
+    ein.store("ein_1000.pkl")
+    print(" done ***\n")
+
+    ###############################################################################
+    # Continue simulation by another 1s.
+    print("\n*** Continuing simulation ***")
+    nest.Simulate(T_sim)
+    dplot.add_to_plot(ein, lbl="Continued simuation", t_min=T_sim, t_max=2*T_sim)
```
Comment (YounesBouhadjar) [inline pynest/examples/store_restore_network.py:314]: typo in simulation

File: pynest/examples/store_restore_network.py [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+
+    ###############################################################################
+    # Clear kernel, restore network from file and simulate for 1s.
+    print("\n*** Reloading and resuming simulation ***")
+    nest.ResetKernel()
+    nest.SetKernelStatus({"local_num_threads": 4})
+    ein2 = EINetwork()
+    ein2.restore("ein_1000.pkl")
+    nest.Simulate(T_sim)
+    dplot.add_to_plot(ein2, lbl="Reloaded simuation")
```
Comment (YounesBouhadjar) [inline pynest/examples/store_restore_network.py:324]: typo in simulation

File: pynest/examples/store_restore_network.py [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+    ###############################################################################
+    # Repeat previous step. This shall result in *exactly* the same results as
+    # the previous run because we use the same random seed.
+    print("\n*** Reloading and resuming simulation (same seed) ***")
+    nest.ResetKernel()
+    nest.SetKernelStatus({"local_num_threads": 4})
+    ein2 = EINetwork()
+    ein2.restore("ein_1000.pkl")
+    nest.Simulate(T_sim)
+    dplot.add_to_plot(ein2, lbl="Reloaded simuation (same seed)")
```
Comment (YounesBouhadjar) [inline pynest/examples/store_restore_network.py:335]: typo in simulation

File: pynest/examples/store_restore_network.py [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+
+            Second row, right: Simulating for 1000ms after loading the stored network
+            into a clean kernel (green). Time runs from 0ms and only connectivity, V_m and
+            synaptic weights are restored. Dynamics differ somewhat from continuation.
+
+            Third row, right: Same as in second row with identical random seed (red),
+            resulting in identical spike patterns.
+
+            Forth row, right: Simulating for 1000ms from same stored network state as
+            above but with different random seed yields differen spike patterns (purple).
```
Comment (YounesBouhadjar) [inline pynest/examples/store_restore_network.py:240]: typo in the second different

File: pynest/examples/store_restore_network.py [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+
+        ###############################################################################
+        # Build dictionary with relevant network information:
+        #   - membrane potential for all neurons in each population
+        #   - source, target and weight of all connections
+        # Dictionary entries are Pandas Dataframes.
+        #
+        # Strictly speaking, we would not need to store the weight of the inhibitory
+        # synapses since they are fixed, but we do so out of symmetry and to make it
+        # easier to add plasticity for inihibitory connections later.
```
Comment (YounesBouhadjar) [inline pynest/examples/store_restore_network.py:137]: typo in inhibitory

**Review by YounesBouhadjar — CHANGES_REQUESTED** [review]
Thanks @heplesser, these are some comments:

**Comment by YounesBouhadjar** [comment]
> I have now revised the text in the figure
The text in the caption is good now

**Comment by heplesser** [comment]
@YounesBouhadjar Thanks for tracking all my typos, they should be fixed now. Do you know the suggestion feature in Github code review? Mark one or several lines, then open a comment and click the rectangle with ± in its menu and enter your suggested change. The author of the PR can then directly accept your suggestion.

**Review by clinssen — APPROVED** [review]

**Comment by YounesBouhadjar** [comment]
> @YounesBouhadjar Thanks for tracking all my typos, they should be fixed now. Do you know the suggestion feature in Github code review? Mark one or several lines, then open a comment and click the rectangle with ± in its menu and enter your suggested change. The author of the PR can then directly accept your suggestion.

Yes, I know it, I just missed seeing the ± in the menu

**Review by YounesBouhadjar — APPROVED** [review]

**Pull request merged**